File: check-deps.pl
#!/usr/bin/env perl

use strict;
use warnings;
use version;
use feature 'say';
use Pod::Usage;

# Configuration
use constant {
    CONAN_REMOTE => 'conancenter',
    CONANFILE    => 'conanfile.txt',
};

sub extract_dependencies {
    open my $fh, '<', CONANFILE or die "Unable to open @{[CONANFILE]}: $!\n";

    my $in_requires = 0;
    my @deps;

    while (my $line = <$fh>) {
        chomp $line;

        # check for section headers
        if ($line =~ /^\[requires\]/) {
            $in_requires = 1;
            next;
        }

        $in_requires = 0 if $line =~ /^\[/;
        next unless $in_requires;

        $line =~ s/#.*$//;        # remove comments
        $line =~ s/^\s+|\s+$//g;  # trim whitespace
        next unless $line;        # skip empty lines

        push @deps, $line;
    }

    close $fh;
    return @deps;
}

sub latest_version_from {
    my ($dep_name, $output) = @_;
    my @versions;

    for my $line (split /\n/, $output) {
        if ($line =~ /^\s*\Q$dep_name\E\/([0-9A-Za-z_.+\-]+)/) {
            push @versions, version->parse($1);
        }
    }

    return unless @versions;
    @versions = sort { $a <=> $b } @versions;
    return $versions[-1];
}

sub check_dependency {
    my ($entry) = @_;

    my ($dep_name, $rest) = split m{/}, $entry, 2;

    unless (defined $rest && length $dep_name && length $rest) {
        say "⚠ Unable to parse dependency entry: $entry";
        return;
    }

    my ($current_version) = split /@/, $rest, 2;
    $current_version = version->parse($current_version);

    unless (defined $current_version && length $current_version) {
        say "⚠ Unable to determine current version for $entry";
        return;
    }

    my $output = `conan search $dep_name -r @{[CONAN_REMOTE]} 2>&1`;

    if ($? != 0) {
        printf "⚠ %s search failed (current %s)\n", $dep_name, $current_version;
        say "\n--- Conan Output ---\n$output\n--- End Conan Output ---";
        return;
    }

    my $latest_version = latest_version_from($dep_name, $output);

    unless (defined $latest_version) {
        printf "⚠ %s unable to determine latest version (current %s)\n", $dep_name, $current_version;
        return;
    }

    if ($current_version == $latest_version) {
        printf "✓ %s is up to date (version %s)\n", $dep_name, $current_version;
    } else {
        printf "⚠ %s update available (%s → %s)\n", $dep_name, $current_version, $latest_version;
    }
}

sub main {
    pod2usage(-verbose => 2, -exitval => 0) if @ARGV && $ARGV[0] eq '--help';

    say "Checking for Conan dependency updates...";

    my @dependencies = extract_dependencies();

    unless (@dependencies) {
        say "No dependencies found in @{[CONANFILE]}";
        exit 0;
    }

    for my $dep (@dependencies) {
        check_dependency($dep);
    }
}

main();

__END__

=head1 NAME

check-deps.pl - Check for updated Conan dependencies

=head1 SYNOPSIS

perl check-deps.pl [--help]

=head1 DESCRIPTION

This script checks the conanfile.txt for dependencies and compares their current
versions against the latest versions available in the Conan remote repository.
It reports which dependencies are up-to-date and which have updates available.

=head1 OPTIONS

=over 4

=item B<--help>

Display this help message and exit.

=back

=cut

File: check-deps.py
#!/usr/bin/env python3

import argparse
import re
import subprocess
import sys
from pathlib import Path
from packaging.version import Version, InvalidVersion

# Configuration
CONAN_REMOTE = 'conancenter'
CONANFILE = 'conanfile.py'


def extract_dependencies():
    """Extract dependencies from conanfile.py."""
    conanfile_path = Path(CONANFILE)
    if not conanfile_path.exists():
        print(f"Error: {CONANFILE} not found.")
        return []

    content = conanfile_path.read_text()

    # Match requires = (...) or requires = [...]
    match = re.search(
        r'(?:^|\n)\s*requires\s*=\s*(\((?:.*?)\)|\[(?:.*?)\])',
        content,
        re.DOTALL
    )
    if match:
        requires_block = match.group(1)
        deps = re.findall(r'["\']([^"\']+)["\']', requires_block)
        return deps

    # Match requires = "..."
    match = re.search(
        r'(?:^|\n)\s*requires\s*=\s*["\']([^"\']+)["\']',
        content
    )
    if match:
        return [match.group(1)]

    return []


def latest_version_from(dep_name, output):
    """Parse conan search output to find the latest version."""
    versions = []
    # Pattern to match dep_name/version
    # Perl regex: /^\s*\Q$dep_name\E\/([0-9A-Za-z_.+\-]+)/
    pattern = re.compile(
        rf'^\s*{re.escape(dep_name)}/([0-9A-Za-z_.+\-]+)',
        re.MULTILINE
    )

    for match in pattern.finditer(output):
        v_str = match.group(1)
        try:
            versions.append(Version(v_str))
        except InvalidVersion:
            # Fallback for non-PEP 440 versions if necessary,
            # but packaging.version is requested.
            continue

    if not versions:
        return None

    return sorted(versions)[-1]


def check_dependency(entry):
    """Check a single dependency for updates."""
    parts = entry.split('/', 1)
    if len(parts) != 2:
        print(f"⚠ Unable to parse dependency entry: {entry}")
        return

    dep_name, rest = parts
    current_version_str = rest.split('@', 1)[0]

    try:
        current_version = Version(current_version_str)
    except InvalidVersion:
        print(f"⚠ Unable to determine current version for {entry}")
        return

    try:
        # Run conan search
        result = subprocess.run(
            ['conan', 'search', dep_name, '-r', CONAN_REMOTE],
            capture_output=True,
            text=True,
            check=False
        )
        output = result.stdout + result.stderr
    except Exception as e:
        print(f"⚠ Error running conan search for {dep_name}: {e}")
        return

    if result.returncode != 0:
        print(f"⚠ {dep_name} search failed (current {current_version})")
        print("\n--- Conan Output ---")
        print(output)
        print("--- End Conan Output ---")
        return

    latest_version = latest_version_from(dep_name, output)

    if latest_version is None:
        print(
            f"⚠ {dep_name} unable to determine latest version "
            f"(current {current_version})"
        )
        return

    if current_version == latest_version:
        print(f"✓ {dep_name} is up to date (version {current_version})")
    elif current_version < latest_version:
        print(
            f"⚠ {dep_name} update available "
            f"({current_version} → {latest_version})"
        )
    else:
        # current_version > latest_version
        print(
            f"✓ {dep_name} is up to date (version {current_version} "
            f"is newer than remote {latest_version})"
        )


def main():
    parser = argparse.ArgumentParser(
        description="Check for updated Conan dependencies"
    )
    parser.parse_args()

    print("Checking for Conan dependency updates...")

    dependencies = extract_dependencies()

    if not dependencies:
        print(f"No dependencies found in {CONANFILE}")
        sys.exit(0)

    for dep in dependencies:
        check_dependency(dep)


if __name__ == "__main__":
    main()
